### PR TITLE
8255678: Add Mutex::try_lock version without rank checks

### PR DESCRIPTION
**Pull request by pchilano**
Please review the following patch to add method Mutex::try_lock_without_rank_check().

The ranking system is used to prevent deadlocks while trying to acquire ownership of a Mutex/Monitor. It does so by preventing circular waits. For that we assign a rank to each Mutex/Monitor and we enforce that ownership is acquired in order of decreasing rank. Today we validate this order everytime we acquire ownership of a Mutex/Monitor. We can relax this validation for try_lock() in cases where failure to acquire ownership will not block the thread's progress, i.e. cases where the action can just be skipped (e.g. see JDK-8254877). For those cases, even if the Mutex/Monitor is out of rank order the thread will not block if it already has an owner, so circular waits are not possible.

The patch adds method Mutex::try_lock_without_rank_check() which behaves just like the normal try_lock() but skips rank checks. If the call is successful the method sets field _skip_rank_check for that Mutex/Monitor, so that future rank validations (in case the thread acquires another Mutex/Monitor) will not fail when traversing the list of already acquired locks and checking they are in increasing rank order. The field is reset back on unlock(). For Monitors we still validate that wait() is done on the lock with less rank from all the owned locks.

Today rank checks are executed after locking the underlying PlatformMutex/PlatforMonitor. This means that we might end up in an actual deadlock before asserting it. For the wait() case there is a similar situation, where we check that a thread waits on the Monitor with less rank from all the locks it owns after the underlying PlatformMonitor wait() call returns. Since doing those checks there might be already too late I moved them to the beginning of the respective lock/wait methods. 

Also, all these rank checks are skipped under certain conditions (Mutex/Monitor of rank native or suspend_resume, or at safepoint) and I didn't change any of that. I realized there are inconsistencies though in skipping the rank check when acquiring a Mutex/Monitor of rank suspend_resume but not when traversing the list of owned locks and checking that those are in increasing rank order.

I noticed the rank related fields are defined within an "#ifndef PRODUCT" macro (optimized and debug) but the code that uses them is defined within "#ifdef ASSERT" macros (debug only), so I did a little clean up there too. Basically from that non PRODUCT block I moved everything to an ASSERT block, except for _allow_vm_block, and consolidated it with the ASSERT block defined at the end of the class definition.

In summary the patch contains the following changes:
- Add method Mutex::try_lock_without_rank_check()
- Move rank checks before trying to lock/wait on the underlying PlatformMonitor
- Clean up definitions for optimized/debug builds
- Add a new gtest to test for different rank violations scenarios

Tested it in mach5 tier1-6. I also checked it builds for release/optimized/fastdebug cases.

Thanks,
Patricio

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8255678](https://bugs.openjdk.java.net/browse/JDK-8255678): Add Mutex::try_lock version without rank checks

### Reviewers
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to c17a9c1c8253cb90e3f963269fb25a9d2325e73d

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1242/head:pull/1242`
`$ git checkout pull/1242`

### CODE REVIEW

File: src/hotspot/share/runtime/mutex.cpp
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/mutex.cpp:421]: nit - missing blank - s/this->rank(),/this->rank(), /

File: src/hotspot/share/runtime/mutex.hpp
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/mutex.hpp:91]: nit - typo: s/whend/when/

**Review by dcubed-ojdk — APPROVED** [review]
Your call on whether to fix the nits. I like the new test!
Most of the longer stress test runs happen in Tier7 and
Tier8 so you should also run those.

File: src/hotspot/share/runtime/mutex.cpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp]: Fixed.

File: src/hotspot/share/runtime/mutex.hpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.hpp]: Fixed.

**Comment by pchilano** [comment]
> Your call on whether to fix the nits. I like the new test!
> Most of the longer stress test runs happen in Tier7 and
> Tier8 so you should also run those.
Sounds good, I'll give it a run on tiers 7 and 8 then. 
Thanks for reviewing this Dan!

Patricio

File: src/hotspot/share/runtime/mutex.cpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/mutex.cpp:161]: This is being fixed by https://github.com/openjdk/jdk/pull/1247

File: src/hotspot/share/runtime/mutex.cpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/mutex.cpp:220]: Not clear why this is necessary here, and why it is a guarantee rather than assert ??

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Patricio,
This looks good to me, modulo the overlap with JDK-8256383 (the try_lock bug).
One query below.
Thanks,
David

**Comment by pchilano** [comment]
> Hi Patricio,
> This looks good to me, modulo the overlap with JDK-8256383 (the try_lock bug).
> One query below.
> Thanks,
> David
Hi David,

Thanks for looking at this.

File: src/hotspot/share/runtime/mutex.cpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp:161]: I'll wait until you push and then I'll merge. I still have to check the owner here though, because if the thread already owns the lock, calling check_rank() from try_lock will do the wrong check (it will assume it's being called from wait()). An alternative would be to add a boolean parameter to check_rank(), set only from try_lock() so I can differentiate the two calls.

File: src/hotspot/share/runtime/mutex.cpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp:220]: I looked at the history of this check. The guarantee has been since day1 on the hg repo as "self->is_Java_thread()" and then changed in 8226228 to "self->is_active_Java_thread()". I moved it at the beginning of wait() since it seems the right place to check it just after retrieving self. It doesn't make sense to keep it as a guarantee when every other check is an assert so I will change it.

File: src/hotspot/share/runtime/mutex.cpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp:161]: How about just:
```
bool Mutex::try_lock() {
  Thread * const self = Thread::current();
  DEBUG_ONLY(if (!owned_by_self()) check_rank();)
  return try_lock_inner(self);
}
```
keeping try_lock_without_rank_check() as is. Then I can merge try_lock_inner() with your patch.

Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp]: Ok, makes sense. I updated the code with the suggested change.

Thanks!

File: src/hotspot/share/runtime/mutex.cpp
Comment (coleenp) [inline src/hotspot/share/runtime/mutex.cpp:405]: Wouldn't you want to test check_can_be_skipped and skip all of this?

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp
Comment (coleenp) [inline test/hotspot/gtest/runtime/test_mutex_rank.cpp:88]: I think you should add a test that _skip_rank_check is reset, ie. lock mutex_rankA_plus_one with plain try_lock and make sure it tests the rank.

**Review by coleenp — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/mutex.cpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp:405]: For the wait() case, today we always check that the thread doesn't wait while holding a Mutex/Monitor of rank less than or equal to special, and I didn't want to change that.
But I could do it for the "lock()/lock_without_safepoint_check()/try_lock()" case, and return before calling get_least_ranked_lock().

File: test/hotspot/gtest/runtime/test_mutex_rank.cpp
Comment (pchilano) [reply test/hotspot/gtest/runtime/test_mutex_rank.cpp:88]: Sounds good, I'll add that.

**Review by dholmes-ora — APPROVED** [review]
A couple of minor nits but otherwise I'm good with this.
Thanks,
David

File: src/hotspot/share/runtime/mutex.cpp
Comment (coleenp) [reply src/hotspot/share/runtime/mutex.cpp:405]: too many conditions!  Now I see it.  Okay.

**Review by coleenp — APPROVED** [review]
Looks great. Thanks!

File: src/hotspot/share/runtime/mutex.cpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp]: Fixed.

File: src/hotspot/share/runtime/mutex.cpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp]: Fixed.

File: src/hotspot/share/runtime/mutex.cpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.cpp]: Right, I just changed it to check the owner directly.

File: src/hotspot/share/runtime/mutex.hpp
Comment (pchilano) [reply src/hotspot/share/runtime/mutex.hpp]: I removed set_owner(Thread* owner).

Comment (pchilano) [reply src/hotspot/share/runtime/mutex.hpp]: s/set_owner/owned_by_self

**Review by dholmes-ora — APPROVED** [review]
LGTM!

Thanks,
David

**Review by dcubed-ojdk — APPROVED** [review]
Still good.

**Comment by pchilano** [comment]
Thanks @dcubed-ojdk, @dholmes-ora and @coleenp for the reviews!

**Pull request closed without merging**
